OpenRA.Mods.Common/Traits/Production.cs

```diff
@@ -120,25 +120,62 @@ public virtual void DoProduction(Actor self, ActorInfo producee, ExitInfo exitin
 
 				var notifyOthers = self.World.ActorsWithTrait<INotifyOtherProduction>();
 				foreach (var notify in notifyOthers)
-					notify.Trait.UnitProducedByOther(notify.Actor, self, newUnit);
+					notify.Trait.UnitProducedByOther(notify.Actor, self, newUnit, type);
 
 				foreach (var t in newUnit.TraitsImplementing<INotifyBuildComplete>())
 					t.BuildingComplete(newUnit);
 			});
 		}
 
-		public virtual bool Produce(Actor self, ActorInfo producee, string factionVariant)
+		protected virtual ExitInfo SelectExit(Actor self, ActorInfo producee, string type, Func<ExitInfo, bool> p)
+		{
+			if (type != null)
+			{
+				var exit = self.RandomExitOrDefault(type, p);
+				if (exit != null)
+					return exit;
+			}
+
+			var bi = producee.TraitInfoOrDefault<BuildableInfo>();
+			if (bi != null)
+			{
+				if (Info.Produces.Contains(bi.BuildAtProductionType))
```

Comment:
This is where my problem starts - this tag was made up to resolve the kennel-barracks issue in RA and thereby only used in `ClassicProductionQueue`s, so I'm unsure if this will work as intended in residual `ProductionQueue` mechanics. (I guess that can be tested by defining a common `BuildAtProductionType: Infantry` on all TD infantry and then setting the exits to simple `Infantry` as well, but I haven't made that test yet.)

Although now that I think about it, getting the above working doesn't sound much of a feature creeping.